### PR TITLE
CORE-422: In Core Crypto, Define NetworkPeer

### PR DESCRIPTION
**Pull request by EBGToo**
@michael-brd have a look at this.  I 'flowed' the peer into BRSyncManager; modified cryptoWalletMangerConnect() with a NULLABLE peer argument.

### CODE REVIEW

File: crypto/BRCryptoPeer.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+cryptoPeerCreate (BRCryptoNetwork network,
+                  const char *address,
+                  uint16_t port,
+                  const char *publicKey) {
+    // Require `address`
+    if (NULL == address || 0 == strlen (address)) return NULL;
+
+    // Require `address` to parse as a INET address
+    struct in_addr addr;
+    if (-1 == ascii2addr (AF_INET, address, &addr)) return NULL;
```
Comment (michael-brd) [inline crypto/BRCryptoPeer.c]: How does IPv6 (which seems to be supported by the `BRPeerManager`) factor?

File: crypto/BRCryptoWalletManager.c [hunk 2]
```diff
@@ -99,6 +100,7 @@ cryptoWalletManagerCreateInternal (BRCryptoCWMListener listener,
     cwm->state   = cryptoWalletManagerStateInit (CRYPTO_WALLET_MANAGER_STATE_CREATED);
     cwm->addressScheme = scheme;
     cwm->path = strdup (path);
+    cwm->peer = NULL;
```
Comment (michael-brd) [inline crypto/BRCryptoWalletManager.c]: Presumably don't need this on the CWM given that it is in the `connect` call?

**Comment by michael-brd** [comment]
@EBGToo, looks good to me. Obviously a PoC/WiP so just added some comments on questions rather than a proper review. Nice work!

File: crypto/BRCryptoWalletManager.c (hunk 2 above)
Comment (EBGToo) [reply crypto/BRCryptoWalletManager.c]: Yes, that is true.  Although a comment in `connect()` notes that if we kept the peer we can avoid calling BRWalletManagerSetFixedPeer() every time.

The possible down side to calling SetFixedPeer is that it always does an explicit disconnect - which might be an issue?

**Comment by EBGToo** [comment]
I extended the Core Demo Swift app to demonstrate an optional BTC peer.  Worked!

File: crypto/BRCryptoWalletManager.c (hunk 2 above)
Comment (michael-brd) [reply crypto/BRCryptoWalletManager.c]: Well, with mode changes, we'd have to think about where we actually want to cache this (if at all).

File: crypto/BRCryptoPeer.c (hunk 1 above)
Comment (michael-brd) [reply crypto/BRCryptoPeer.c]: Also, not sure `ascii2addr` is portable.

Comment (EBGToo) [reply crypto/BRCryptoPeer.c]: CORE-578

File: crypto/BRCryptoWalletManager.c (hunk 2 above)
Comment (EBGToo) [reply crypto/BRCryptoWalletManager.c]: I removed `cwm->peer`

File: Swift/BRCrypto/BRCryptoNetwork.swift [hunk 3]
```diff
@@ -292,3 +311,60 @@ public final class NetworkFee: Equatable {
...
+    
+//    internal convenience init? (network: BRCryptoNetwork, serialization: Data) {
+//        var bytes = [UInt8](serialization)
+//        guard let core = cryptoPeerCreateFromSerialization (network, &bytes, bytes.count)
+//            else { return nil }
+//
+//        self.init (core: core, take: false)
+//    }
+
+    public static func == (lhs: NetworkPeer, rhs: NetworkPeer) -> Bool {
```
Comment (michael-brd) [inline Swift/BRCrypto/BRCryptoNetwork.swift]: Do we need an `IsIdentical()` at the C level?

File: Swift/BRCrypto/BRCryptoNetwork.swift [hunk 4]
```diff
@@ -292,3 +311,60 @@ public final class NetworkFee: Equatable {
...
+    public let publicKey: String?
+
+//    /// The serialization to allow reconstructing this peer
+//    public var serialization: Data {
+//        var bytesCount: Int = 0
+//        let bytes = cryptoPeerSerialize(core, &bytesCount)
+//        return Data (bytes: bytes!, count: bytesCount)
+//    }
+
+    internal init (core: BRCryptoPeer, take: Bool) {
```
Comment (michael-brd) [inline Swift/BRCrypto/BRCryptoNetwork.swift]: Do we need a `cryptoPeerGive()`?

File: Swift/BRCryptoTests/BRCryptoWalletTests.swift [hunk 5]
```diff
@@ -165,9 +165,10 @@ class BRCryptoWalletTests: BRCryptoSystemBaseTests {
              // EventMatcher (event: WalletManagerEvent.syncProgress(timestamp: nil, percentComplete: 0), strict: false),
              EventMatcher (event: WalletManagerEvent.walletChanged(wallet: wallet), strict: true, scan: true),
              
-             EventMatcher (event: WalletManagerEvent.syncEnded(error: nil), strict: false, scan: true),
+             EventMatcher (event: WalletManagerEvent.syncEnded(reason: WalletManagerSyncStoppedReason.complete), strict: false, scan: true),
```
Comment (michael-brd) [inline Swift/BRCryptoTests/BRCryptoWalletTests.swift]: 😨 ...sorry...

File: Swift/BRCryptoTests/BRCryptoWalletManagerTests.swift [hunk 6]
```diff
@@ -127,9 +127,10 @@ class BRCryptoWalletManagerTests: BRCryptoSystemBaseTests {
              // We might not see `syncProgress`
              // EventMatcher (event: WalletManagerEvent.syncProgress(timestamp: nil, percentComplete: 0), strict: false),
 
-             EventMatcher (event: WalletManagerEvent.syncEnded(error: nil), strict: false, scan: true),
+             EventMatcher (event: WalletManagerEvent.syncEnded(reason: WalletManagerSyncStoppedReason.complete), strict: false, scan: true),
```
Comment (michael-brd) [inline Swift/BRCryptoTests/BRCryptoWalletManagerTests.swift]: 😨 ... sorry here too...

File: Swift/BRCryptoTests/BRCryptoTransferTests.swift [hunk 7]
```diff
@@ -171,9 +171,10 @@ class BRCryptoTransferTests: BRCryptoSystemBaseTests {
              // EventMatcher (event: WalletManagerEvent.syncProgress(timestamp: nil, percentComplete: 0), strict: false),
              EventMatcher (event: WalletManagerEvent.walletChanged(wallet: wallet), strict: true, scan: true),
 
-             EventMatcher (event: WalletManagerEvent.syncEnded(error: nil), strict: false, scan: true),
+             EventMatcher (event: WalletManagerEvent.syncEnded(reason: WalletManagerSyncStoppedReason.requested), strict: false, scan: true),
```
Comment (michael-brd) [inline Swift/BRCryptoTests/BRCryptoTransferTests.swift]: 😨 ... now I am really sorry...

**Comment by michael-brd** [comment]
@EBGToo, I added the Java implementation. I was seeing connect *attempts* but not successes to "103.99.168.100:8333". Did the demo app successfully connect?

**Comment by michael-brd** [comment]
Did you see a successful connect (in the iOS demo app)? I can't get one (on Android as well as iOS).

File: Swift/BRCrypto/BRCryptoNetwork.swift (hunk 3 above)
Comment (michael-brd) [reply Swift/BRCrypto/BRCryptoNetwork.swift]: I don't know that we do (for Java I did the same as you); just a question.

File: Swift/BRCrypto/BRCryptoWalletManager.swift [hunk 8]
```diff
@@ -133,12 +133,17 @@ public final class WalletManager: Equatable, CustomStringConvertible {
...
-    //    var walletFactory: WalletFactory { get set }
-
-    /// Connect to network and begin managing wallets for account
-    public func connect () {
-        cryptoWalletManagerConnect (core)
+    ///
+    /// Connect to the network and begin managing wallets.
+    ///
+    /// - Parameter peer: An optional NetworkPeer to use on the P2P network.  It is unusual to
+    ///     provide a peer as P2P networks will dynamically descover suitable peers.
```
Comment (ewerx) [inline Swift/BRCrypto/BRCryptoWalletManager.swift]: discover (sp)

File: Swift/BRCrypto/BRCryptoNetwork.swift (hunk 3 above)
Comment (EBGToo) [reply Swift/BRCrypto/BRCryptoNetwork.swift]: Yes, I'll update.

File: Swift/BRCrypto/BRCryptoWalletManager.swift (hunk 8 above)
Comment (EBGToo) [reply Swift/BRCrypto/BRCryptoWalletManager.swift]: Fixed.

File: crypto/testCrypto.c [hunk 9]
```diff
@@ -824,7 +824,7 @@ runCryptoWalletManagerLifecycleTest (BRCryptoAccount account,
        BRCryptoWallet wallet = cryptoWalletManagerGetWallet (manager);
 
        // connect, disconnect
-       cryptoWalletManagerConnect (manager);
+       cryptoWalletManagerConnect (manager, NULL);
```
Comment (EBGToo) [inline crypto/testCrypto.c]: Here is where I'm sorry...

File: Swift/BRCrypto/BRCryptoNetwork.swift (hunk 3 above)
Comment (EBGToo) [reply Swift/BRCrypto/BRCryptoNetwork.swift]: Perhaps needs a Java update, for consistency.

File: Swift/BRCrypto/BRCryptoNetwork.swift (hunk 4 above)
Comment (michael-brd) [reply Swift/BRCrypto/BRCryptoNetwork.swift]: You have no `defer { cryptoPeerGive() }` in the Swift. Is that not required?

File: Swift/BRCrypto/BRCryptoNetwork.swift (hunk 3 above)
Comment (michael-brd) [reply Swift/BRCrypto/BRCryptoNetwork.swift]: Yes, pushing shortly.

File: Swift/BRCrypto/BRCryptoNetwork.swift (hunk 4 above)
Comment (EBGToo) [reply Swift/BRCrypto/BRCryptoNetwork.swift]: Done.

**Comment by EBGToo** [comment]
Hmmm, the JIRA number and the branch name have been inconsistent the whole time...

**Pull request merged**